tsfc/coffee_mode.py

```diff
@@ -0,0 +1,282 @@
+                atomic_index[atomic] = next(index)
+            connection.append(atomic_index[atomic])
+        connections.append(tuple(connection))
+
+    if len(atomic_index) == 0:
+        return ()
+    if len(atomic_index) == 1:
+        optimal_atomics, = iterkeys(atomic_index)
+        return (optimal_atomics, )
+
+    # set up the ILP
+    import pulp as ilp
+    ilp_prob = ilp.LpProblem('gem factorise', ilp.LpMinimize)
+    ilp_var = ilp.LpVariable.dicts('node', range(len(atomic_index)), 0, 1, ilp.LpBinary)
+
+    # Objective function
+    # Minimise number of factors to pull. If same number, favour factor with larger extent
+    big = 1e20  # some arbitrary big number
+    ilp_prob += ilp.lpSum(ilp_var[index] * (big - index_extent(atomic, argument_indices))
+                          for atomic, index in iteritems(atomic_index))
+
+    # constraints
+    for connection in connections:
+        ilp_prob += ilp.lpSum(ilp_var[index] for index in connection) >= 1
+
+    ilp_prob.solve()
+    if ilp_prob.status != 1:
+        raise RuntimeError("Something bad happened during ILP")
+
+    def optimal(atomic):
+        return ilp_var[atomic_index[atomic]].value() == 1
+
+    optimal_atomics = filter(optimal, iterkeys(atomic_index))
+    return tuple(optimal_atomics)
+
+
+def factorise_atomics(monomials, optimal_atomics, argument_indices):
+    """Group and factorise monomials using a list of atomics as common
+    subexpressions. Create new monomials for each group and optimise them recursively.
+
+    :arg monomials: an iterable of :class:`Monomial`s, all of which should have
+                    the same sum indices
+    :arg optimal_atomics: list of tuples of atomics to be used as common subexpression
+    :arg argument_indices: tuple of argument indices
+
+    :returns: an iterable of factorised :class:`Monomials`s, or the original
+              input if no changes are made.
+    """
+    if not optimal_atomics or len(monomials) < 2:
+        return monomials
+
+    # Group monomials with respect to each optimal atomic
+    def group_key(monomial):
+        for oa in optimal_atomics:
+            if oa in monomial.atomics:
+                return oa
+        assert False, "Expect at least one optimal atomic per monomial."
+    factor_group = groupby(monomials, key=group_key)
+
+    # We should not drop monomials
+    assert sum(len(ms) for _, ms in factor_group) == len(monomials)
+
+    sum_indices = next(iter(monomials)).sum_indices
+    new_monomials = []
+    for oa, monomials in factor_group:
+        # Create new MonomialSum for the factorised out terms
+        sub_monomials = []
+        for monomial in monomials:
+            atomics = list(monomial.atomics)
+            atomics.remove(oa)  # remove common factor
+            sub_monomials.append(Monomial((), tuple(atomics), monomial.rest))
+        sub_monomials = optimise_monomials(sub_monomials, argument_indices)
+        assert len(sub_monomials) > 0
```

Comment:
Actually it's probably okay since we don't simplify `stuff + (-1)*stuff`, and if we do (later), we will just get Zero out of `monomial_sum_to_expression`, so we still get a monomial with Zero as `rest`, and propagate upwards.
So I think I just need to remove this assertion. Will test it.